sandbox/store.go

```diff
@@ -0,0 +1,66 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package sandbox
+
+import (
+	"context"
+	"time"
+
+	"github.com/gogo/protobuf/types"
+)
+
+// Sandbox is an object stored in metadata database
+type Sandbox struct {
+	// ID uniquely identifies the sandbox in a namespace
+	ID string
+	// Labels provide metadata extension for a sandbox
+	Labels map[string]string
+	// Runtime shim to use for this sandbox
+	Runtime RuntimeOpts
+	// Spec carries the runtime specification used to implement the sandbox
+	Spec *types.Any
```

Comment:
In the exported interface I wonder if we can just stick to `interface{}` and when we need a `*types.Any` we can type assert for that or marshal to it if something else. I think it is already going to be a challenge using types from `gogo/protobuf` so it is better we find a better solution now.

Reply:
This is relatively straightforward to do (mostly adds `typeurl.MarshalAny/typeurl.UnmarshalAny` back and forth). However one particular limitation of this is that containerd needs to know the type to be passed in advance (via `typeurl.Register`), so custom types will not be allowed without forking contained. This is fine as we currently use container spec and its already registered, but might be a limitation if someone decides to pass custom spec to the runtime.

Possibly migration to protobuf's native Any type would be a better option? ([1](https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/any.proto) [2](https://github.com/golang/protobuf/blob/master/ptypes/any.go))